WIP: [CSL-2383] Extend testing infrastructure

edsko opened the pull request:
When done this PR will pave the way to hook up @uroboros 's work on the wallet to the inductive wallet generator. Indeed, a large part of this PR will consist of a reworked version of the work that @uroboros has already done.

edsko commented:
>  I think that it is more correct to ignore epoch-boundary blocks: they are only convenience devices, not semantic

Sure, that's fair, but the very first block _does_ point to the "genesis block" as its predecessor. So conceivably something odd happens at epoch boundaries too. Not disagreeing with you btw, but I just wanted to make sure :)

edsko commented:
Yessir :)

uroboros commented on wallet-new/test/unit/Wallet/Abstract/Cardano.hs:
```diff
@@ -69,44 +72,47 @@ interpretT :: forall h e m. (Monad m, Hash h Addr)
 -------------------------------------------------------------------------------}
 
-equivalentT :: forall h m. (Hash h Addr, MonadIO m)
+equivalentT :: forall h m. (HasConfiguration, Hash h Addr, MonadIO m)
```
HasConfiguration provides the context for the call to Cardano.Core 'selectOwnAddresses' (called from equivalentT)

edsko commented:
@uroboros CI is failing due to warnings in the `Kernel` module. Would you mind fixing those, so that we can merge this PR? Thanks! :)

parsonsmatt commented on wallet-new/test/unit/Wallet/Abstract/Cardano.hs:
```diff
@@ -0,0 +1,272 @@
+    -- This is ugly, but we only discover the bootstrap transaction once we
+    -- descend down the 'Inductive' wallet. We will 'put' the right context
+    -- before the first call to 'int'.
+    runIntT (error "initialized later") (fst <$> go ind'')
```
Can we embed the comment into the error message? eg `error "You called 'int' before the context was initialized."`

edsko replied:
Changed it to `interpretT: the impossible happened`. Explaining the full setup isn't very useful here, I think, but at least this is a lot more greppable if/when it would ever arise.

parsonsmatt commented on wallet-new/test/unit/Wallet/Abstract/Cardano.hs:
```diff
@@ -0,0 +1,272 @@
+    )
+    equivalenceNotCheckedName
+    equivalenceNotCheckedReason
+    equivalenceNotCheckedInductive
```
4 space indent

parsonsmatt commented on wallet-new/test/unit/WalletUnitTest.hs:
```diff
@@ -177,44 +183,106 @@ testPureWallet = do
+    choosePoorActor transCtxt' = do
+        let poorActors = Map.elems . actorsPoor . tcActors $ transCtxt'
+        poorIx <- choose (0, length poorActors -1)
+        let esk =  encKpEnc . poorKey $ poorActors !! poorIx
```
If the intent is to pick a random element from `poorActors`, you can do `poorActor <- elements poorActors`

edsko commented on wallet-new/src/Cardano/Wallet/Kernel.hs:
```diff
@@ -34,34 +120,256 @@ import Pos.Core (TxAux)
+-- | Modify Wallet State with given modifier function
+modifyWalletState :: PassiveWallet -> WalletId -> (State -> State) -> IO State
+modifyWalletState pw wid modifyF = do
+    w <- fromJust <$> findWallet pw wid
```
As [mentioned](https://github.com/input-output-hk/cardano-sl/pull/2779#pullrequestreview-114279891), this doesn't matter as this is just a placeholder for now. The acid-state design provides detailed error messages for the various error cases.

edsko commented on wallet-new/src/Cardano/Wallet/Kernel.hs:
```diff
@@ -34,34 +120,256 @@ import Pos.Core (TxAux)
+            -> IO ()
+applyBlock' pw b wid = do
+    (State utxo' pending' balance') <- getWalletState pw wid
+    w <- fromJust <$> findWallet pw wid
```
This is temporary code.

edsko commented on wallet-new/src/Cardano/Wallet/Kernel/Types.hs:
```diff
@@ -0,0 +1,117 @@
+-- | (Unsigned block) containing resolved transactions
+data ResolvedBlock = ResolvedBlock {
+      -- | Transactions in the block
+      rbTxs :: [ResolvedTx]
```
No, it's fine to have a block without any transactions.

edsko replied:
Will also change `ResolvedTxInputs` accordingly.

edsko commented on wallet-new/src/Cardano/Wallet/WalletLayer.hs:
```diff
@@ -24,18 +24,19 @@ import qualified Cardano.Wallet.WalletLayer.Legacy as Legacy
-
 bracketKernelPassiveWallet
-    :: forall m n a. (MonadMask n, Monad m)
+    :: forall m n a. (HasConfiguration, MonadIO m, MonadIO n, MonadMask n, Monad m)
```
Sadly, it's almost impossible to call into the Cardano core without having that `HasConfiguration` thing available.

edsko commented on wallet-new/src/Cardano/Wallet/Kernel/Types.hs:
```diff
@@ -0,0 +1,117 @@
+-- | Signed transaction along with its resolved inputs
+--
+-- Invariant: number of inputs @==@ number of resolved inputs
+type RawResolvedTx = (TxAux, ResolvedTxInputs)
```
Ok.

edsko commented on wallet-new/src/Cardano/Wallet/Kernel/Types.hs:
```diff
@@ -0,0 +1,117 @@
+-- | Signed block along with its resolved inputs
+--
+-- Invariant: number of transactions @==@ number of resolved transaction inputs
+type RawResolvedBlock = (MainBlock, ResolvedBlockInputs)
```
Ok.

edsko commented on wallet-new/src/Cardano/Wallet/Kernel/Types.hs:
```diff
@@ -0,0 +1,117 @@
+  The raw types are the original types along with some additional information.
+  In the derived types (below) we actually lose the original types (and
+  therefore signatures etc.).
+-------------------------------------------------------------------------------}
```
Will move to the definition of `ResolvedTx` and `ResolvedBlock`.

edsko commented on wallet-new/src/Cardano/Wallet/Kernel/Types.hs:
```diff
@@ -0,0 +1,117 @@
+-- | (Unsigned) transaction with inputs resolved
+data ResolvedTx = ResolvedTx {
+      -- | Transaction inputs
+      rtxInputs  :: [(TxIn, ResolvedInput)]
```
Ok.

edsko commented:
Addressed review comments as indicated in the comments above, with the exception of nitpicks on formatting issues in @uroboros 's code. I don't necessarily disagree with them but I don't think it's the best use of my time right now :)

edsko commented:
Rebased on `develop` in the hope of making CI go green.
